disnake/ext/commands/common_bot_base.py

```diff
@@ -574,18 +575,151 @@ def reload_extension(self, name: str, *, package: Optional[str] = None) -> None:
+    ) -> List[str]:
+        """
+        Finds all extensions in a given module, also traversing into sub-packages.
+
+        See :ref:`ext_commands_extensions_load` for details on how packages are found.
+
+        .. versionadded:: 2.7
+
+        .. note::
+            This imports all *packages* (not all modules) in the given path(s)
+            to access the ``__path__`` attribute for finding submodules,
+            unless they are filtered by the ``ignore`` parameter.
+
+        Parameters
+        ----------
+        root_module: :class:`str`
+            The module/package name to search in, for example ``cogs.admin``.
+            Also supports paths in the current working directory.
+        package: Optional[:class:`str`]
+            The package name to resolve relative imports with.
+            This is required when ``root_module`` is a relative module name, e.g ``.cogs.admin``.
+            Defaults to ``None``.
+        ignore: Optional[Union[Iterable[:class:`str`], Callable[[:class:`str`], :class:`bool`]]]
+            An iterable of module names to ignore, or a callable that's used for ignoring
+            modules (where the callable returning ``True`` results in the module being ignored).
+            Defaults to ``None``, i.e. no modules are ignored.
+
+            If it's an iterable, module names that start with any of the given strings will be ignored.
+
+        Raises
+        ------
+        ExtensionError
+            The given root module could not be found,
+            or the name of the root module could not be resolved using the provided ``package`` parameter.
+        ValueError
+            ``root_module`` is a path and outside of the cwd.
+        TypeError
+            The ``ignore`` parameter is of an invalid type.
+        ImportError
+            A package couldn't be imported.
+
+        Returns
+        -------
+        List[:class:`str`]
+            The list of full extension names.
+        """
+        if "/" in root_module or "\\" in root_module:
+            path = os.path.relpath(root_module)
+            if ".." in path:
+                raise ValueError(
+                    "Paths outside the cwd are not supported. Try using the module name instead."
+                )
+            root_module = path.replace(os.sep, ".")
+
+        # `find_spec` already calls `resolve_name`, but we want our custom error handling here
+        root_module = self._resolve_name(root_module, package)
+
+        if not (spec := importlib.util.find_spec(root_module)):
+            raise errors.ExtensionError(
+                f"Unable to find root module '{root_module}'", name=root_module
+            )
+
+        if not (paths := spec.submodule_search_locations):
+            raise errors.ExtensionError(
+                f"Module '{root_module}' is not a package", name=root_module
+            )
+
+        return list(disnake.utils._walk_modules(paths, prefix=f"{spec.name}.", ignore=ignore))
```

Comment:
Why the cast to list?

Reply:
I see. I usually cast to `tuple` if the intent is to consume/create state independent iterable. Obviously negligible in terms of performance, but I think it feels nicer.

Reply:
Fair point, returning an immutable type here makes sense. Changed it to tuple in https://github.com/DisnakeDev/disnake/pull/796/commits/bece2cf86ac16782f54470e1e7eb9a10726bb682.